// @smithy/core/dist-es/submodules/serde/index.js
export { copyDocumentWithTransform } from "./copyDocumentWithTransform";
export { dateToUtcString, parseRfc3339DateTime, parseRfc3339DateTimeWithOffset, parseRfc7231DateTime, parseEpochTimestamp, } from "./date-utils";
export { generateIdempotencyToken } from "./generateIdempotencyToken";
export { LazyJsonString } from "./lazy-json";
export { logger, parseBoolean, expectBoolean, expectNumber, expectFloat32, expectInt, expectInt32, expectShort, expectByte, expectNonNull, expectObject, expectString, expectUnion, expectLong, strictParseDouble, strictParseFloat, strictParseFloat32, strictParseLong, strictParseInt, strictParseInt32, strictParseShort, strictParseByte, limitedParseDouble, handleFloat, limitedParseFloat, limitedParseFloat32, } from "./parse-utils";
export { quoteHeader } from "./quote-header";
export { _parseEpochTimestamp, _parseRfc3339DateTimeWithOffset, _parseRfc7231DateTime, } from "./schema-serde-lib/schema-date-utils";
export { splitEvery } from "./split-every";
export { splitHeader } from "./split-header";
export { NumericValue, nv } from "./value/NumericValue";
export { fromHex, toHex } from "./util-hex-encoding/hex-encoding";
export { fromBase64 } from "./util-base64/fromBase64";
export { toBase64 } from "./util-base64/toBase64";
export { calculateBodyLength } from "./util-body-length/calculateBodyLength";
export { fromUtf8 } from "./util-utf8/fromUtf8";
export { toUint8Array } from "./util-utf8/toUint8Array";
export { toUtf8 } from "./util-utf8/toUtf8";
export { fromArrayBuffer, fromString } from "./util-buffer-from/buffer-from";
export { isArrayBuffer } from "./is-array-buffer/is-array-buffer";
export { deserializerMiddleware } from "./middleware-serde/deserializerMiddleware";
export { deserializerMiddlewareOption, serializerMiddlewareOption, getSerdePlugin, } from "./middleware-serde/serdePlugin";
export { serializerMiddleware } from "./middleware-serde/serializerMiddleware";
export { Hash } from "./hash-node/hash-node";
export { Uint8ArrayBlobAdapter } from "./util-stream/blob/Uint8ArrayBlobAdapter";
export { ChecksumStream } from "./util-stream/checksum/ChecksumStream";
export { createChecksumStream } from "./util-stream/checksum/createChecksumStream";
export { createBufferedReadable } from "./util-stream/createBufferedReadable";
export { getAwsChunkedEncodingStream } from "./util-stream/getAwsChunkedEncodingStream";
export { headStream } from "./util-stream/headStream";
export { sdkStreamMixin } from "./util-stream/sdk-stream-mixin";
export { splitStream } from "./util-stream/splitStream";
export { isReadableStream, isBlob } from "./util-stream/stream-type-check";
export { v4 } from "./uuid/v4";
