// @aws-sdk/client-rds-data/dist-es/models/errors.js
import { RDSDataServiceException as __BaseException } from "./RDSDataServiceException";
export class AccessDeniedException extends __BaseException {
    name = "AccessDeniedException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "AccessDeniedException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AccessDeniedException.prototype);
    }
}
export class BadRequestException extends __BaseException {
    name = "BadRequestException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "BadRequestException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, BadRequestException.prototype);
    }
}
export class DatabaseErrorException extends __BaseException {
    name = "DatabaseErrorException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DatabaseErrorException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DatabaseErrorException.prototype);
    }
}
export class DatabaseNotFoundException extends __BaseException {
    name = "DatabaseNotFoundException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DatabaseNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DatabaseNotFoundException.prototype);
    }
}
export class DatabaseResumingException extends __BaseException {
    name = "DatabaseResumingException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "DatabaseResumingException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DatabaseResumingException.prototype);
    }
}
export class DatabaseUnavailableException extends __BaseException {
    name = "DatabaseUnavailableException";
    $fault = "server";
    constructor(opts) {
        super({
            name: "DatabaseUnavailableException",
            $fault: "server",
            ...opts,
        });
        Object.setPrototypeOf(this, DatabaseUnavailableException.prototype);
    }
}
export class ForbiddenException extends __BaseException {
    name = "ForbiddenException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ForbiddenException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ForbiddenException.prototype);
    }
}
export class HttpEndpointNotEnabledException extends __BaseException {
    name = "HttpEndpointNotEnabledException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "HttpEndpointNotEnabledException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, HttpEndpointNotEnabledException.prototype);
    }
}
export class InternalServerErrorException extends __BaseException {
    name = "InternalServerErrorException";
    $fault = "server";
    constructor(opts) {
        super({
            name: "InternalServerErrorException",
            $fault: "server",
            ...opts,
        });
        Object.setPrototypeOf(this, InternalServerErrorException.prototype);
    }
}
export class InvalidResourceStateException extends __BaseException {
    name = "InvalidResourceStateException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidResourceStateException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidResourceStateException.prototype);
    }
}
export class InvalidSecretException extends __BaseException {
    name = "InvalidSecretException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidSecretException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidSecretException.prototype);
    }
}
export class SecretsErrorException extends __BaseException {
    name = "SecretsErrorException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "SecretsErrorException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, SecretsErrorException.prototype);
    }
}
export class ServiceUnavailableError extends __BaseException {
    name = "ServiceUnavailableError";
    $fault = "server";
    constructor(opts) {
        super({
            name: "ServiceUnavailableError",
            $fault: "server",
            ...opts,
        });
        Object.setPrototypeOf(this, ServiceUnavailableError.prototype);
    }
}
export class StatementTimeoutException extends __BaseException {
    name = "StatementTimeoutException";
    $fault = "client";
    dbConnectionId;
    constructor(opts) {
        super({
            name: "StatementTimeoutException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, StatementTimeoutException.prototype);
        this.dbConnectionId = opts.dbConnectionId;
    }
}
export class TransactionNotFoundException extends __BaseException {
    name = "TransactionNotFoundException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "TransactionNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, TransactionNotFoundException.prototype);
    }
}
export class NotFoundException extends __BaseException {
    name = "NotFoundException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "NotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, NotFoundException.prototype);
    }
}
export class UnsupportedResultException extends __BaseException {
    name = "UnsupportedResultException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "UnsupportedResultException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, UnsupportedResultException.prototype);
    }
}
